In [1]:
import pandas as pd
from scipy import stats

In [2]:
property_data = pd.read_csv('https://raw.githubusercontent.com/edyoda/Data-Scientist-program/master/Practice%20Problems/data/property%20data.csv')
property_data

PID  ST_NUM     ST_NAME OWN_OCCUPIED NUM_BEDROOMS NUM_BATH SQ_FT
0  100001000.0   104.0      PUTNAM            Y            3        1  1000
1  100002000.0   197.0   LEXINGTON            N            3      1.5    --
2  100003000.0     NaN   LEXINGTON            N          NaN        1   850
3  100004000.0   201.0    BERKELEY           12            1      NaN   700
4          NaN   203.0    BERKELEY            Y            3        2  1600
5  100006000.0   207.0    BERKELEY            Y          NaN        1   800
6  100007000.0     NaN  WASHINGTON          NaN            2   HURLEY   950
7  100008000.0   213.0     TREMONT            Y            1        2   NaN
8  100009000.0   215.0     TREMONT            Y           na        2  1800

In [3]:
missing_values = ['na','HURLEY','--','12','n/a','1.5']
property_data = pd.read_csv('https://raw.githubusercontent.com/edyoda/Data-Scientist-program/master/Practice%20Problems/data/property%20data.csv', na_values = missing_values)
property_data

PID  ST_NUM     ST_NAME OWN_OCCUPIED  NUM_BEDROOMS  NUM_BATH  \
0  100001000.0   104.0      PUTNAM            Y           3.0       1.0   
1  100002000.0   197.0   LEXINGTON            N           3.0       NaN   
2  100003000.0     NaN   LEXINGTON            N           NaN       1.0   
3  100004000.0   201.0    BERKELEY          NaN           1.0       NaN   
4          NaN   203.0    BERKELEY            Y           3.0       2.0   
5  100006000.0   207.0    BERKELEY            Y           NaN       1.0   
6  100007000.0     NaN  WASHINGTON          NaN           2.0       NaN   
7  100008000.0   213.0     TREMONT            Y           1.0       2.0   
8  100009000.0   215.0     TREMONT            Y           NaN       2.0   

    SQ_FT  
0  1000.0  
1     NaN  
2   850.0  
3   700.0  
4  1600.0  
5   800.0  
6   950.0  
7     NaN  
8  1800.0

Filling the missing value in PID with 100005000.0 since data is given in sequence and 100005000.0 is missing

In [4]:
property_data['PID'].fillna(100005000.0,inplace = True)

In [5]:
property_data

PID  ST_NUM     ST_NAME OWN_OCCUPIED  NUM_BEDROOMS  NUM_BATH  \
0  100001000.0   104.0      PUTNAM            Y           3.0       1.0   
1  100002000.0   197.0   LEXINGTON            N           3.0       NaN   
2  100003000.0     NaN   LEXINGTON            N           NaN       1.0   
3  100004000.0   201.0    BERKELEY          NaN           1.0       NaN   
4  100005000.0   203.0    BERKELEY            Y           3.0       2.0   
5  100006000.0   207.0    BERKELEY            Y           NaN       1.0   
6  100007000.0     NaN  WASHINGTON          NaN           2.0       NaN   
7  100008000.0   213.0     TREMONT            Y           1.0       2.0   
8  100009000.0   215.0     TREMONT            Y           NaN       2.0   

    SQ_FT  
0  1000.0  
1     NaN  
2   850.0  
3   700.0  
4  1600.0  
5   800.0  
6   950.0  
7     NaN  
8  1800.0

filling missing value if ST_NUM of LEXINGTON as 199.0,since street numbers of same ST_NAME vary by 2
and WASHINGTON by 210.0

In [6]:
property_data[2:3]['ST_NUM'] = 199.0
property_data[6:7]['ST_NUM'] = 210.0

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [7]:
property_data

PID  ST_NUM     ST_NAME OWN_OCCUPIED  NUM_BEDROOMS  NUM_BATH  \
0  100001000.0   104.0      PUTNAM            Y           3.0       1.0   
1  100002000.0   197.0   LEXINGTON            N           3.0       NaN   
2  100003000.0   199.0   LEXINGTON            N           NaN       1.0   
3  100004000.0   201.0    BERKELEY          NaN           1.0       NaN   
4  100005000.0   203.0    BERKELEY            Y           3.0       2.0   
5  100006000.0   207.0    BERKELEY            Y           NaN       1.0   
6  100007000.0   210.0  WASHINGTON          NaN           2.0       NaN   
7  100008000.0   213.0     TREMONT            Y           1.0       2.0   
8  100009000.0   215.0     TREMONT            Y           NaN       2.0   

    SQ_FT  
0  1000.0  
1     NaN  
2   850.0  
3   700.0  
4  1600.0  
5   800.0  
6   950.0  
7     NaN  
8  1800.0

since in BERKELEY most of the houses are own occupied,filling the NaN cell as Y 
and for washington filling it with the mode of the column which is Y

In [8]:
property_data[3:4]['OWN_OCCUPIED'] = 'Y'

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [9]:
property_data['OWN_OCCUPIED'].fillna(property_data['OWN_OCCUPIED'].mode()[0],inplace = True)
property_data

PID  ST_NUM     ST_NAME OWN_OCCUPIED  NUM_BEDROOMS  NUM_BATH  \
0  100001000.0   104.0      PUTNAM            Y           3.0       1.0   
1  100002000.0   197.0   LEXINGTON            N           3.0       NaN   
2  100003000.0   199.0   LEXINGTON            N           NaN       1.0   
3  100004000.0   201.0    BERKELEY            Y           1.0       NaN   
4  100005000.0   203.0    BERKELEY            Y           3.0       2.0   
5  100006000.0   207.0    BERKELEY            Y           NaN       1.0   
6  100007000.0   210.0  WASHINGTON            Y           2.0       NaN   
7  100008000.0   213.0     TREMONT            Y           1.0       2.0   
8  100009000.0   215.0     TREMONT            Y           NaN       2.0   

    SQ_FT  
0  1000.0  
1     NaN  
2   850.0  
3   700.0  
4  1600.0  
5   800.0  
6   950.0  
7     NaN  
8  1800.0

In [10]:
# in data NUM_BATH for PID = 100002000.0 is 1.5 which might be a data entry error,rpalcong it with 1.0 and the SQ_FT as 1000
property_data[1:2]['NUM_BATH'] = 1.0
property_data[1:2]['SQ_FT'] = 1000.0


# in data for PID = 100003000.0 is NUM_BATH = 1 and SQ_FT = 850, from observation guessing NUM_BEDROOMS = 2
property_data[2:3]['NUM_BEDROOMS'] = 2.0

# in data for PID = 100004000.0 is NUM_BEDROOMS = 1 and SQ_FT = 700, from observation guessing NUM_BATH = 1.for 700SQ_FT
property_data[3:4]['NUM_BATH'] = 1.0

# in data for PID = 100006000.0 is  NUM_BATH = 1 and SQ_FT = 800, from observation guessing NUM_BEDROOMS = 1.for 800SQ_FT
property_data[5:6]['NUM_BEDROOMS'] = 1.0

# in data for PID = 100007000.0 is  NUM_BEDROOMS = 2 and SQ_FT = 950, from observation guessing NUM_BATH = 2.for 950SQ_FT
property_data[6:7]['NUM_BATH'] = 2.0

# in data for PID = 100008000.0 is  NUM_BEDROOMS = 1 and NUM_BATH = 2, from observation guessing SQ_FT = 900
property_data[7:8]['SQ_FT'] = 900.0

# in data for PID = 100009000.0 is  NUM_BATH = 2 and SQ_FT = 1800, from observation guessing NUM_BEDROOMS = 3.for 1800SQ_FT
property_data[8:]['NUM_BEDROOMS'] = 3.0


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = val

In [11]:
property_data

PID  ST_NUM     ST_NAME OWN_OCCUPIED  NUM_BEDROOMS  NUM_BATH  \
0  100001000.0   104.0      PUTNAM            Y           3.0       1.0   
1  100002000.0   197.0   LEXINGTON            N           3.0       1.0   
2  100003000.0   199.0   LEXINGTON            N           2.0       1.0   
3  100004000.0   201.0    BERKELEY            Y           1.0       1.0   
4  100005000.0   203.0    BERKELEY            Y           3.0       2.0   
5  100006000.0   207.0    BERKELEY            Y           1.0       1.0   
6  100007000.0   210.0  WASHINGTON            Y           2.0       2.0   
7  100008000.0   213.0     TREMONT            Y           1.0       2.0   
8  100009000.0   215.0     TREMONT            Y           3.0       2.0   

    SQ_FT  
0  1000.0  
1  1000.0  
2   850.0  
3   700.0  
4  1600.0  
5   800.0  
6   950.0  
7   900.0  
8  1800.0

In [12]:
property_data.dtypes

PID             float64
ST_NUM          float64
ST_NAME          object
OWN_OCCUPIED     object
NUM_BEDROOMS    float64
NUM_BATH        float64
SQ_FT           float64
dtype: object

# Chi-Square test

In [13]:
table = pd.crosstab(property_data['ST_NAME'],property_data['OWN_OCCUPIED'])
table

OWN_OCCUPIED  N  Y
ST_NAME           
BERKELEY      0  3
LEXINGTON     2  0
PUTNAM        0  1
TREMONT       0  2
WASHINGTON    0  1

In [14]:
from scipy.stats import chi2_contingency
chi2_contingency(table)

(8.999999999999998, 0.061099480960332735, 4, array([[0.66666667, 2.33333333],
        [0.44444444, 1.55555556],
        [0.22222222, 0.77777778],
        [0.44444444, 1.55555556],
        [0.22222222, 0.77777778]]))

p =  0.061099480960332735 which is greater than 0.05 hence, there is no asssociation between two variable
but most the cell data is < 5 hence Chi-Square Test is not reliable

# Using researchpy library

# Chi-Square Test

In [15]:
import researchpy as rp

In [16]:
table, results = rp.crosstab(property_data['ST_NAME'],property_data['OWN_OCCUPIED'],test='chi-square')
    
print(table)
results

           OWN_OCCUPIED       
                      N  Y All
ST_NAME                       
BERKELEY              0  3   3
LEXINGTON             2  0   2
PUTNAM                0  1   1
TREMONT               0  2   2
WASHINGTON            0  1   1
All                   2  7   9


Chi-square test  results
0  Pearson Chi-square ( 4.0) =    9.0000
1                    p-value =    0.0611
2                 Cramer's V =    1.0000

# Fisher's Exact Test

In [17]:
table, results = rp.crosstab(property_data['ST_NAME'],property_data['OWN_OCCUPIED'],test='fisher')
    
print(table)
results

ValueError: The input `table` must be of shape (2, 2).

in jupyter fisher exact test can not be performed for sizes greater than (2,2)

# g-test

In [18]:
table, results = rp.crosstab(property_data['ST_NAME'],property_data['OWN_OCCUPIED'],test='g-test')
    
print(table)
results

           OWN_OCCUPIED       
                      N  Y All
ST_NAME                       
BERKELEY              0  3   3
LEXINGTON             2  0   2
PUTNAM                0  1   1
TREMONT               0  2   2
WASHINGTON            0  1   1
All                   2  7   9


G-test  results
0  Log-likelihood ratio ( 4.0) =    9.5347
1                      p-value =    0.0490
2                   Cramer's V =    1.0293

p-value = 0.0490
hence according to g-test the two variables are related to each other